src/npe2/_plugin_manager.py

```diff
@@ -358,14 +362,36 @@ def register(
             self._npe1_adapters.append(manifest)
         else:
             self._contrib.index_contributions(manifest)
+            self._populate_command_menu_map(manifest)
         self.events.plugins_registered.emit({manifest})
 
+    def _populate_command_menu_map(self, manifest: PluginManifest):
+        self._command_menu_map[manifest.name] = defaultdict(dict)
+        for command in manifest.contributions.commands or ():
+            # command IDs are keys in map
+            # each value is a dict menu_id: list of MenuCommands
+            # for the command and menu
+            associated = self._get_associated_menus(manifest, command.id)
+            self._command_menu_map[manifest.name][command.id] = associated
+
+    def _get_associated_menus(
+        self, manifest: PluginManifest, command_id: str
+    ) -> Dict[str, List[MenuCommand]]:
+        menus = manifest.contributions.menus or dict()
+        associated_menus = defaultdict(list)
+        for menu_id, items in menus.items():
+            for item in items:
+                if getattr(item, "command", "") == command_id:
+                    associated_menus[menu_id].append(item)
+        return associated_menus
```

Comment:
Ah this is backwards. You should be able to do this in a single pass of the menus:
- iterate over the menus, but
- map commands back to menus
- since you use a defaultdict it's ok if some menu-less commands are missed, but you could add them by keeping a set of already-seen commands and then putting in the missed ones at the end.
 
So:

```suggestion
        menu_map = self._command_menu_map[manifest.name]  # just for conciseness below
        for menu_id, menu_items in manifest.contributions.menus or ():
            # command IDs are keys in map
            # each value is a dict menu_id: list of MenuCommands
            # for the command and menu
            for item in menu_items:
                if command_id := getattr(item, "command", None) is not None:
                    menu_map[command_id].append(item)
```